src/display/annotation_layer.js

```diff
@@ -2366,18 +2393,70 @@ class AnnotationLayer {
    * @memberof AnnotationLayer
    */
   static update(parameters) {
-    const transform = `matrix(${parameters.viewport.transform.join(",")})`;
+    const { page, viewport, annotations, annotationCanvasMap, div } =
+      parameters;
+    const transform = viewport.transform;
+    const matrix = `matrix(${transform.join(",")})`;
+
+    let scale, ownMatrix;
+    if (annotations.some(annotation => annotation.hasOwnCanvas)) {
```

Comment:
Rather than having to iterate through the annotations *twice*, given that most pages won't have any canvases, can't you simply calculate this "special" matrix lazily the first time it's need in the loop below instead?